File: Computer_Science/Algorithm/Learning/Algorithms/回溯/组合型/code/39.组合总和.cpp
/*
 * @lc app=leetcode.cn id=39 lang=cpp
 *
 * [39] 组合总和
 */

#include <ratio>
#include<vector>
#include<algorithm>
using namespace std;
// @lc code=start
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> path;
    
    void dfs(vector<int>& candidates, int i, int target) {
        int n = candidates.size();
        if (i == n) return;
        if (target < 0) return;
        if (target == 0) {
            ans.emplace_back(path);
            return;
        }

        for (int j = i; j < n; j++) {
            path.push_back(candidates[j]);
            // 还是从j开始表明下次还可以选这个元素
            dfs(candidates, j, target - candidates[j]);
            path.pop_back();
        }
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        dfs(candidates, 0, target);
        return ans;
    }
};
// @lc code=end

File: Computer_Science/Algorithm/leetcode/hot100/39.组合总和.java
/*
 * @lc app=leetcode.cn id=39 lang=java
 *
 * [39] 组合总和
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> path = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        dfs(0, candidates, target, path, ans);
        return ans;
    }

    /**
     * 子集型题目, 选和不选
     * 不过这里可以重复选同一个数字
     * @param i
     * @param candidates
     * @param target
     */
    private void dfs(int i, int[] candidates, int target, List<Integer> path, List<List<Integer>> ans) {
        if (target == 0) { // 找到了答案
            ans.add(new ArrayList<>(path));
            return;
        }
        if (i >= candidates.length) return;

        int x = candidates[i];
        // 选当前元素
        if (target - x >= 0) {
            path.add(x);
            dfs(i, candidates, target-x, path, ans);
            path.remove(path.size()-1);
        }

        // 不选当前元素
        dfs(i+1, candidates, target, path, ans);
    }

}
// @lc code=end

